renku/core/session/notebook_service.py

```diff
@@ -0,0 +1,386 @@
+        _, is_user_registered = _get_token(renku_url=self._renku_url())
+        return is_user_registered
+
+    def _auth_header(self) -> Dict[str, str]:
+        """Get the authentication header with the JWT token or cookie needed to authenticate with Renku."""
+        if self._is_user_registered():
+            return {"Authorization": f"Bearer {self._token()}"}
+        return {"Cookie": f"anon-id={self._token()}"}
+
+    def _wait_for_session_status(
+        self,
+        name: Optional[str],
+        status: str,
+    ):
+        if not name:
+            return
+        now = datetime.now()
+        timeout_time = now + timedelta(seconds=self.DEFAULT_TIMEOUT_SECONDS)
+        while now < timeout_time:
+            res = self._send_renku_request(
+                "get", f"{self._notebooks_url()}/servers/{name}", headers=self._auth_header()
+            )
+            if res.status_code == 404 and status == "stopping":
+                return
+            if res.status_code == 200 and status != "stopping":
+                if res.json().get("status", {}).get("state") == status:
+                    return
+            sleep(5)
+        raise errors.NotebookServiceSessionError(f"Waiting for the session {name} to reach status {status} timed out.")
+
+    def _get_in_progress_gitlab_job_id(self, gitlab_project_name: str, commit_sha: str) -> Optional[str]:
+        """Get the ID of any job that is in progress and matches the specified arguments."""
+        project_name_enc = gitlab_project_name.replace("/", "%2F")
+        res = self._send_renku_request(
+            "get",
+            f"{self._renku_url()}/api/projects/{project_name_enc}/jobs",
+            params={"scope": ["running", "pending", "created"]},
+            headers=self._auth_header(),
+        )
+        if res.status_code == 200:
+            for job in res.json():
+                if job.get("commit", {}).get("id") == commit_sha:
+                    return job["id"]
+        return None
+
+    def _wait_for_ci_jobs_completion(
+        self,
+        gitlab_project_name: str,
+        commit_sha: str,
+    ):
+        """Wait for all in-progress CI jobs related to the arguments to finish."""
+        if not self._is_user_registered():
+            return
+        now = datetime.now()
+        timeout_time = now + timedelta(seconds=self.DEFAULT_TIMEOUT_SECONDS)
+        while now < timeout_time:
+            if self._get_in_progress_gitlab_job_id(gitlab_project_name, commit_sha) is None:
+                return
+            sleep(5)
+        raise errors.NotebookServiceSessionError(
+            f"Waiting for the CI jobs for project {gitlab_project_name} to finish timed out."
+        )
+
+    def _commit_and_push_checks(self):
+        client = _get_client()
+        remote = get_remote(client.repository)
+        if self._is_user_registered():
+            if client.repository.is_dirty() or len(client.repository.untracked_files) > 0:
```

Comment:
```suggestion
            if client.repository.is_dirty(untracked_files=True):
```

Reply:
cool did not know about this

Reply:
I did the change my self in a batch with a bunch of others.